File: MedicationTimer.cs
// MedicationTimer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

class Medication
{
    [JsonPropertyName("id")] public int Id { get; set; }
    [JsonPropertyName("name")] public string Name { get; set; }
    [JsonPropertyName("dosage")] public string Dosage { get; set; }
    [JsonPropertyName("interval_hours")] public double IntervalHours { get; set; }
    [JsonPropertyName("last_taken")] public string LastTaken { get; set; }
    [JsonPropertyName("history")] public List<string> History { get; set; } = new List<string>();

    public bool IsDue()
    {
        var last = DateTime.Parse(LastTaken);
        var next = last.AddHours(IntervalHours);
        return DateTime.Now >= next;
    }

    public DateTime NextDose()
    {
        var last = DateTime.Parse(LastTaken);
        return last.AddHours(IntervalHours);
    }

    public void TakeDose()
    {
        var now = DateTime.Now;
        LastTaken = now.ToString("o");
        History.Add(now.ToString("o"));
    }
}

class TimerApp
{
    private List<Medication> medications = new List<Medication>();
    private readonly string dataFile = "medications.json";
    private readonly JsonSerializerOptions options = new JsonSerializerOptions { WriteIndented = true };

    public TimerApp() { Load(); }

    private void Load()
    {
        if (!File.Exists(dataFile)) return;
        string json = File.ReadAllText(dataFile);
        medications = JsonSerializer.Deserialize<List<Medication>>(json) ?? new List<Medication>();
    }

    private void Save()
    {
        string json = JsonSerializer.Serialize(medications, options);
        File.WriteAllText(dataFile, json);
    }

    public void AddMedication(string name, string dosage, double interval)
    {
        int maxId = medications.Any() ? medications.Max(m => m.Id) : 0;
        var med = new Medication
        {
            Id = maxId + 1,
            Name = name,
            Dosage = dosage,
            IntervalHours = interval,
            LastTaken = DateTime.Now.ToString("o")
        };
        medications.Add(med);
        Save();
        Console.WriteLine($"💊 Added medication: {name} {dosage} every {interval}h (ID: {med.Id})");
    }

    public void List()
    {
        if (!medications.Any()) { Console.WriteLine("No medications."); return; }
        Console.WriteLine("\n📋 Medications:");
        foreach (var m in medications)
        {
            string due = m.IsDue() ? "\u001b[31m🔴 DUE\u001b[0m" :
                        $"\u001b[32m⏳ {m.NextDose():HH:mm}\u001b[0m";
            Console.WriteLine($"  [{m.Id}] {m.Name} {m.Dosage} – {due}");
        }
    }

    public void TakeDose(int id)
    {
        var m = medications.FirstOrDefault(med => med.Id == id);
        if (m == null) { Console.WriteLine($"❌ Medication with ID {id} not found."); return; }
        m.TakeDose();
        Save();
        Console.WriteLine($"✅ Took {m.Name} {m.Dosage} at {DateTime.Now:HH:mm}");
    }

    public void CheckDue()
    {
        var due = medications.Where(m => m.IsDue()).ToList();
        if (!due.Any()) { Console.WriteLine("✅ No medications due right now."); return; }
        Console.WriteLine("\n⏰ Due Medications:");
        foreach (var m in due)
        {
            long overdue = (long)(DateTime.Now - m.NextDose()).TotalMinutes;
            Console.WriteLine($"  [{m.Id}] {m.Name} {m.Dosage} – overdue by {overdue} min");
        }
    }

    public void History()
    {
        if (!medications.Any()) { Console.WriteLine("No medications."); return; }
        Console.WriteLine("\n📜 History (last 5 doses per medication):");
        foreach (var m in medications)
        {
            Console.WriteLine($"  {m.Name}:");
            var recent = m.History.TakeLast(5).ToList();
            if (!recent.Any())
                Console.WriteLine("    (no history)");
            else
                foreach (var ts in recent)
                    Console.WriteLine($"    - {DateTime.Parse(ts):yyyy-MM-dd HH:mm}");
        }
    }

    static void Main(string[] args)
    {
        if (args.Length < 1) { Console.WriteLine("Usage: MedicationTimer <command> [options]"); return; }
        var app = new TimerApp();
        string cmd = args[0];
        switch (cmd)
        {
            case "add":
                if (args.Length < 4) { Console.WriteLine("add <name> <dosage> <interval>"); return; }
                app.AddMedication(args[1], args[2], double.Parse(args[3]));
                break;
            case "list":
                app.List();
                break;
            case "take":
                if (args.Length < 2) { Console.WriteLine("take <id>"); return; }
                app.TakeDose(int.Parse(args[1]));
                break;
            case "check":
                app.CheckDue();
                break;
            case "history":
                app.History();
                break;
            default:
                Console.WriteLine("Unknown command.");
                break;
        }
    }
}

File: MedicationTimer.java
// MedicationTimer.java
import java.io.*;
import java.nio.file.*;
import java.time.*;
import java.time.format.*;
import java.util.*;
import com.google.gson.*;

class Medication {
    int id;
    String name;
    String dosage;
    double intervalHours;
    String lastTaken;
    List<String> history;

    public Medication() {}

    public Medication(int id, String name, String dosage, double intervalHours) {
        this.id = id;
        this.name = name;
        this.dosage = dosage;
        this.intervalHours = intervalHours;
        this.lastTaken = Instant.now().toString();
        this.history = new ArrayList<>();
    }

    public boolean isDue() {
        Instant last = Instant.parse(lastTaken);
        Instant next = last.plusSeconds((long)(intervalHours * 3600));
        return Instant.now().compareTo(next) >= 0;
    }

    public Instant nextDose() {
        Instant last = Instant.parse(lastTaken);
        return last.plusSeconds((long)(intervalHours * 3600));
    }

    public void takeDose() {
        Instant now = Instant.now();
        lastTaken = now.toString();
        history.add(now.toString());
    }
}

class TimerApp {
    private List<Medication> medications = new ArrayList<>();
    private final String dataFile = "medications.json";
    private final Gson gson = new GsonBuilder().setPrettyPrinting().create();

    public TimerApp() { load(); }

    private void load() {
        try {
            Path path = Paths.get(dataFile);
            if (Files.exists(path)) {
                String json = new String(Files.readAllBytes(path));
                Medication[] arr = gson.fromJson(json, Medication[].class);
                medications = Arrays.asList(arr);
            }
        } catch (Exception e) {}
    }

    private void save() {
        try {
            Files.write(Paths.get(dataFile), gson.toJson(medications).getBytes());
        } catch (Exception e) {}
    }

    public void addMedication(String name, String dosage, double interval) {
        int maxId = medications.stream().mapToInt(m -> m.id).max().orElse(0);
        Medication m = new Medication(maxId + 1, name, dosage, interval);
        medications.add(m);
        save();
        System.out.printf("💊 Added medication: %s %s every %.1fh (ID: %d)%n", name, dosage, interval, m.id);
    }

    public void list() {
        if (medications.isEmpty()) {
            System.out.println("No medications.");
            return;
        }
        System.out.println("\n📋 Medications:");
        for (Medication m : medications) {
            String due = m.isDue() ? "\u001B[31m🔴 DUE\u001B[0m" :
                    "\u001B[32m⏳ " + LocalDateTime.ofInstant(m.nextDose(), ZoneId.systemDefault()).format(DateTimeFormatter.ofPattern("HH:mm")) + "\u001B[0m";
            System.out.printf("  [%d] %s %s – %s%n", m.id, m.name, m.dosage, due);
        }
    }

    public void takeDose(int id) {
        for (Medication m : medications) {
            if (m.id == id) {
                m.takeDose();
                save();
                System.out.printf("✅ Took %s %s at %s%n", m.name, m.dosage,
                        LocalDateTime.now().format(DateTimeFormatter.ofPattern("HH:mm")));
                return;
            }
        }
        System.out.printf("❌ Medication with ID %d not found.%n", id);
    }

    public void checkDue() {
        List<Medication> due = new ArrayList<>();
        for (Medication m : medications) if (m.isDue()) due.add(m);
        if (due.isEmpty()) {
            System.out.println("✅ No medications due right now.");
            return;
        }
        System.out.println("\n⏰ Due Medications:");
        for (Medication m : due) {
            long overdue = (Instant.now().getEpochSecond() - m.nextDose().getEpochSecond()) / 60;
            System.out.printf("  [%d] %s %s – overdue by %d min%n", m.id, m.name, m.dosage, overdue);
        }
    }

    public void history() {
        if (medications.isEmpty()) {
            System.out.println("No medications.");
            return;
        }
        System.out.println("\n📜 History (last 5 doses per medication):");
        for (Medication m : medications) {
            System.out.printf("  %s:%n", m.name);
            List<String> recent = m.history;
            if (recent.size() > 5) recent = recent.subList(recent.size() - 5, recent.size());
            if (recent.isEmpty()) {
                System.out.println("    (no history)");
            } else {
                for (String ts : recent) {
                    Instant instant = Instant.parse(ts);
                    System.out.printf("    - %s%n",
                            LocalDateTime.ofInstant(instant, ZoneId.systemDefault()).format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm")));
                }
            }
        }
    }

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.out.println("Usage: MedicationTimer <command> [options]");
            return;
        }
        TimerApp app = new TimerApp();
        String cmd = args[0];
        switch (cmd) {
            case "add":
                if (args.length < 4) { System.out.println("add <name> <dosage> <interval>"); return; }
                app.addMedication(args[1], args[2], Double.parseDouble(args[3]));
                break;
            case "list":
                app.list();
                break;
            case "take":
                if (args.length < 2) { System.out.println("take <id>"); return; }
                app.takeDose(Integer.parseInt(args[1]));
                break;
            case "check":
                app.checkDue();
                break;
            case "history":
                app.history();
                break;
            default:
                System.out.println("Unknown command.");
        }
    }
}
